pkg/operator/resource/resourceapply/storage.go

```diff
@@ -34,6 +34,15 @@ func ApplyStorageClass(client storageclientv1.StorageClassesGetter, recorder eve
 	existingCopy := existing.DeepCopy()
 
 	resourcemerge.EnsureObjectMeta(modified, &existingCopy.ObjectMeta, required.ObjectMeta)
+
+	// Now that we copied everything that matters from required.ObjectMeta, we
+	// should reset required.ObjectMeta so a DeepEqual() comparison is possible
+	required.ObjectMeta = *existingCopy.ObjectMeta.DeepCopy()
```

Comment:
You don't want to mutate your inputs.  I think I see the bug a few lines down: `existingCopy = required.DeepCopy()`.  I see why you did it, but how about doing the required.DeepCopy() up here and then something like this

```go
//storage class doesn't have a spec and we don't want to miss fields.  We have to copy the required to get all fields and then overwrite objectmeta and typemeta from the original
existingCopy := required.DeepCopy()
existingCopy.ObjectMeta = existing.ObjectMeta.DeepCopy()
existingCopy.TypeMeta = existing.TypeMeta.DeepCopy()
```

Reply:
`existingCopy.ObjectMeta = existing.ObjectMeta.DeepCopy()`: here we'd miss the "merged" ObjectMeta from line 36 above.

I tried to incorporate your suggestions the best I could. I think the main reason this code is so confusing is that we need to mutate the objects at each step.